presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushAggregationBelowOuterJoin.java

```diff
@@ -0,0 +1,283 @@
+                        aggregationNode,
+                        joinSource.getCriteria(),
+                        ImmutableList.<Symbol>builder().addAll(rewrittenLeft.getOutputSymbols()).addAll(aggregationNode.getOutputSymbols()).build(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+
+                return coalesceWithNullAggregation(aggregationNode, joinNode);
+            }
+
+            if (joinSource.getType() == JoinNode.Type.RIGHT
+                    && !joinSource.getFilter().isPresent()
+                    && groupsOnAllOuterTableColumns(node, joinSource)
+                    && isDistinct(joinSource.getRight())) {
+                List<Symbol> groupingKeys = joinSource.getCriteria().stream()
+                        .map(JoinNode.EquiJoinClause::getLeft)
+                        .collect(toImmutableList());
+                AggregationNode aggregationNode = new AggregationNode(
+                        node.getId(),
+                        context.defaultRewrite(joinSource.getLeft()),
+                        node.getAggregations(),
+                        node.getFunctions(),
+                        node.getMasks(),
+                        ImmutableList.of(groupingKeys),
+                        node.getStep(),
+                        node.getHashSymbol(),
+                        node.getGroupIdSymbol());
+                PlanNode rewrittenRight = context.defaultRewrite(joinSource.getRight());
+                JoinNode joinNode = new JoinNode(
+                        joinSource.getId(),
+                        joinSource.getType(),
+                        aggregationNode,
+                        rewrittenRight,
+                        joinSource.getCriteria(),
+                        ImmutableList.<Symbol>builder().addAll(aggregationNode.getOutputSymbols()).addAll(rewrittenRight.getOutputSymbols()).build(),
+                        joinSource.getFilter(),
+                        joinSource.getLeftHashSymbol(),
+                        joinSource.getRightHashSymbol());
+                return coalesceWithNullAggregation(aggregationNode, joinNode);
+            }
+            return context.defaultRewrite(node, context.get());
+        }
+
+        // When the aggregation is done after the join, there will be a null value that gets aggregated over
+        // where rows did not exist in the inner table.  For some aggregate functions, such as count, the result
+        // of an aggregation over a single null row is one or zero rather than null. In order to ensure correct results,
+        // we add a coalesce function with the output of the new outer join and the agggregation performed over a single
+        // null row.
+        private PlanNode coalesceWithNullAggregation(AggregationNode aggregationNode, PlanNode outerJoin)
+        {
+            NullLiteral nullLiteral = new NullLiteral();
```

Comment:
I made an attempt-- let me know what you think.  Creating the null aggregation is definitely the messiest part.  I separated that out into its own method, which I think makes everything else much easier to follow.  That bit is still sort of convoluted, but hopefully also somewhat better.